plugins/elantp/fu-elantp-common.h

```diff
@@ -8,6 +8,7 @@
 
 #include <glib.h>
 
+#define ETP_CMD_GET_HID_DESCRIPTOR	0x0001
```

Comment:
This ETP_CMD_GET_HID_DESCRIPTOR can be read from the acpi table or get from the metainfo.xml?
-> Device id : ELANXXXX:XX
-> i2c-address: 0x15
-> HID Descriptor Register Address: 0x0001

Reply:
I've added the DMI check; you can get more hwids for other machines using `sudo fwupdtool hwids`

Reply:
I mean, not all elan tp's "ETP_CMD_GET_HID_DESCRIPTOR" are 0x0001 . 
And this value is defined in ACPI TABLE.

Reply:
> I mean, not all elan tp's "ETP_CMD_GET_HID_DESCRIPTOR" are 0x0001

More information required! :) Is this the kind of thing we should include in the .quirk file like we do the i2c address? Or can you share more details about which ACPI table you mean, and the format of it?

Reply:
I think the point is the ACPI tables should still be populated even if the device is in recovery state, and that information should come from ACPI tables.  It would scale better than duplication in a quirk file.
Can you put this into a recovery state and check what is still exporting from them into `/sys`?

Reply:
> that information should come from ACPI tables

But which table?

Reply:
Generally I'd think DSDT should have it.  Perhaps if you acpidump and decompile and put in a gist we can look it over?

Reply:
So on my system with another vendor's I2C touchpad, it shows up like `/sys/bus/acpi/devices/CUST0001:00`.  I can see in that heirarchy that `physical_node` points back over to the I2C controller:
`../../../../../pci0000:00/0000:00:15.1/i2c_designware.1/i2c-2/i2c-CUST0001:00`

Looking at my DSDT I can see that it's I2C address 0x17, but I don't see the kernel actually exporting that bit into userspace.  @bentiss is there a way that the information requested (i2c address and HID descriptor register address) from an i2c HID export from ACPI entry the kernel parsed to userspace that we're not aware of?

Reply:
cd /sys/bus/acpi/devices/ELANXXXX:XX
cat status
--------------------
sys node also only found i2c address.

Reply:
https://elixir.bootlin.com/linux/v4.5/source/drivers/hid/i2c-hid/i2c-hid.c
wHIDDescRegister -> read from ACPI table.

static int i2c_hid_of_probe(struct i2c_client *client,
		struct i2c_hid_platform_data *pdata)
{
	struct device *dev = &client->dev;
	u32 val;
	int ret;

	ret = of_property_read_u32(dev->of_node, "hid-descr-addr", &val);
	if (ret) {
		dev_err(&client->dev, "HID register address not provided\n");
		return -ENODEV;
	}
	if (val >> 16) {
		dev_err(&client->dev, "Bad HID register address: 0x%08x\n",
			val);
		return -EINVAL;
	}
	pdata->hid_descriptor_address = val;

	return 0;
}

Reply:
static int i2c_hid_acpi_pdata(struct i2c_client *client,
		struct i2c_hid_platform_data *pdata)
{
	static u8 i2c_hid_guid[] = {
		0xF7, 0xF6, 0xDF, 0x3C, 0x67, 0x42, 0x55, 0x45,
		0xAD, 0x05, 0xB3, 0x0A, 0x3D, 0x89, 0x38, 0xDE,
	};
	union acpi_object *obj;
	struct acpi_device *adev;
	acpi_handle handle;
	int ret;

	handle = ACPI_HANDLE(&client->dev);
	if (!handle || acpi_bus_get_device(handle, &adev))
		return -ENODEV;

	obj = acpi_evaluate_dsm_typed(handle, i2c_hid_guid, 1, 1, NULL,
				      ACPI_TYPE_INTEGER);
	if (!obj) {
		dev_err(&client->dev, "device _DSM execution failed\n");
		return -ENODEV;
	}

	pdata->hid_descriptor_address = obj->integer.value;
	ACPI_FREE(obj);

	/* GPIOs are optional */
	ret = acpi_dev_add_driver_gpios(adev, i2c_hid_acpi_gpios);
	return ret < 0 && ret != -ENXIO ? ret : 0;
}

Reply:
That's great, thanks!  I didn't realize that's what the number in status actually exported.  But so in this recovery situation, does that mean that basically the `physical_node` symlink wouldn't exist?

Reply:
It seems that it is difficult to read this information? 
For x86 OS is in the ACPI table, ARM OS is in the DTS.

Reply:
It seems that the I2C device address is covered. Regarding the HID descriptor address, I don't think we export it anywhere in the i2c-hid kernel module.

Reply:
> Regarding the HID descriptor address, I don't think we export it anywhere

But _could_ we? It would save fwupd parsing the ACPI blob itself.

Reply:
> But could we? It would save fwupd parsing the ACPI blob itself.

Depends on what you want/need. My first reaction was to say that we should be able to easily export that from i2c-hid, but if this is a recovery firmware, there is a chance this needs to go lower at the ACPI level. And this gets a little bit more complex, because why would we expose a protocol specific on every ACPI devices?

As Jingle mentioned, the information is not on the ACPI table, but in an ACPI function. So for ACPI devices, this is a matter of just running the blob above, and you'll get the information.

For Arm, and the DTS, well, we are kind of screwed.

Reply:
> but if this is a recovery firmware, there is a chance this needs to go lower at the ACPI level

I guess @jinglewu might know more about this.

> this is a matter of just running the blob above

Reply:
> But can we do this from userspace?

right. Damn. Forgot about that. The answer is "I don't know for sure, but I highly doubt it".

>> For Arm, and the DTS, well, we are kind of screwed.
>
> So perhaps putting it in the fwupd .quirk file isn't so crazy afterall?

Well, there is one thing I am doing more and more, is to use eBPF to monitor unmodified kernel functions. I wonder if we could not instrument the kernel, force a i2c-hid discovery and then fetch all of those information :)

Reply:
> is to use eBPF to monitor unmodified kernel functions

The *whoooosh* as all that goes right over my head :)

> I wonder if we could not instrument the kernel

I think the `we` in that is highly optimistic :) For fwupd, especially for device /recovery/ I'd prefer we KISS if we can help it.

Reply:
Since we're talking about a recovery scenario, what about caching some stuff around this into the history database?

Reply:
Something like `FWUPD_DEVICE_FLAG_IS_RECOVERY`? I could use that for the USB2 VLI hub updates too.

Reply:
I haven't thought through all the details for how it would work yet, but my initial thought is something like if you see an elan touchpad device, create some metadata on that device for the i2c address and the hid descriptor.  Save that in the history database.  Next time elan tries to enumerate if that device isn't showing up, but you see in the history database there was a previous device with that information you use that information to try to probe for recovery scenario.

Reply:
Just to give an idea of what I am talking about:
On fedora, `sudo dnf install bcc`, then run the following snippet as root https://gist.github.com/bentiss/b694253a358c6a2e11da3bc54786a9d2

When `Tracing... Hit Ctrl-C to end.` is shown, use an other terminal to unbind/rebind the wanted i2c-hid device:
```bash
cd /sys/module/i2c_hid/drivers/i2c:i2c_hid
DEVICE=i2c-00304238304C4C44:00 ; echo $DEVICE > unbind ; echo $DEVICE> bind
```

(replace `i2c-00304238304C4C44:00` with the correct I2C device you are interested in)

Note: currently bringing in `bcc` installs llvmlite and do a on the fly compilation. AFAIU, libbpf would be capable of providing kernel agnostic files you can distribute and not need a compiler anymore for the clients. But that's not something I have explored, yet

Reply:
Although super-cool, this isn't really in-scope for fwupd :)

Reply:
> there was a previous device with that information

I think that sounds super-fragile to be honest; I think the simplest thing is just to encode the i2c address in the quirk file. I'm not super keen on making fwupd much more complicated; it's tricky enough to debug as it is.